pytorch_lightning/core/lightning.py

```diff
@@ -1562,6 +1572,16 @@ def optimizer_step(
         """
         optimizer.step(closure=optimizer_closure)
 
+        # TODO: Refactor the following code out of this file.
+        # CAVEAT: If the user overrides this method, then model averaging will be disabled by default.
+        if _TORCH_GREATER_EQUAL_1_10 and self._model_averager is not None:
+            for opt in self.optimizers(use_pl_optimizer=False):
```

Comment:
this hook is getting called per optimizer. we still loop over all optimizers?

Reply:
Yes, this hook is called per optimizer step, and we loop over all the optimizers. Certainly we can cache all the target parameters somewhere and only run such looping once. However, I don't think we should include such optimization in this PR:

1) In practice, a trainer usually only has a very small number of optimizers, so the cost of looping here should be trivial. We probably don't want to do over-optimization unless if necessary, and even if we want, it's better to do this incrementally in a separate PR.
2) More important, for some use cases the number of parameters that compute gradients and get averaged can change over time. For example, some base layers may be frozen gradually during pre-train. In this case, it's much safer to run such loop every time.